src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,277 @@
+        a new KeyCollection for any keys that don't have corresponding
+        ids in results.
+        """
+        unmapped_org_strings: MutableMapping[int, Set[str]] = defaultdict(set)
+        for org_id, strings in keys.mapping.items():
+            for string in strings:
+                try:
+                    self.results[org_id][string]
+                except KeyError:
+                    unmapped_org_strings[org_id].add(string)
+
+        return KeyCollection(unmapped_org_strings)
+
+    def get_mapped_key_strings_to_ints(self) -> MutableMapping[str, int]:
+        """
+        Return the results, but formatted as the following:
+            {
+                "1:a": 10,
+                "1:b": 11,
+                "1:c", 12,
+                "2:e": 13
+            }
+        This is for when we use indexer_cache.set_many()
+        """
+        cache_key_results: MutableMapping[str, int] = {}
+        for org_id, result_dict in self.results.items():
+            for string, id in result_dict.items():
+                key = f"{org_id}:{string}"
+                cache_key_results[key] = id
+
+        return cache_key_results
+
+
+class PGStringIndexerV2(Service):
+    """
+    Provides integer IDs for metric names, tag keys and tag values
+    and the corresponding reverse lookup.
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
+
+    def _get_db_records(self, db_keys: KeyCollection) -> Any:
+        conditions = []
+        for pair in db_keys.as_tuples():
+            organization_id, string = pair
+            conditions.append(Q(organization_id=int(organization_id), string=string))
+
+        query_statement = reduce(or_, conditions)
+
+        return StringIndexerTable.objects.filter(query_statement)
+
+    def bulk_record(
+        self, org_strings: MutableMapping[int, Set[str]]
+    ) -> MutableMapping[int, MutableMapping[str, int]]:
+        """
+        TODO(meredith): add doc string
+        """
+        cache_keys = KeyCollection(org_strings)
+        cache_key_strs = cache_keys.as_strings()
+        cache_results = indexer_cache.get_many(cache_key_strs)
+
+        hits = [k for k, v in cache_results.items() if v is not None]
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "true", "caller": "get_many_ids"},
+            amount=len(hits),
+        )
+        metrics.incr(
+            _INDEXER_CACHE_METRIC,
+            tags={"cache_hit": "false", "caller": "get_many_ids"},
+            amount=len(cache_results) - len(hits),
+        )
+
+        cache_key_results = KeyResults()
+        cache_key_results.add_key_results(
+            [KeyResult.from_string(k, v) for k, v in cache_results.items() if v is not None]
+        )
+
+        mapped_results = cache_key_results.get_mapped_results()
+        db_read_keys = cache_key_results.get_unmapped_keys(cache_keys)
+
+        if not db_read_keys.as_strings():
```

Comment:
you do not need to call the `as_strings()` method here and process all the strings just to check the size of the `KeyCollection`. Please add a `size` method on the `KeyCollection` class that return the size of the number of keys contained.